packages/contracts/contracts/L1/AssetManager.sol

```diff
@@ -0,0 +1,1459 @@
+     *
+     * @return The amount of shares that the Vault would exchange for the amount of assets provided.
+     */
+    function previewDelegate(
+        address validator,
+        uint128 assets
+    ) public view virtual returns (uint128) {
+        return _convertToKroShares(validator, assets);
+    }
+
+    /**
+     * @notice Allows an on-chain or off-chain user to simulate the effects of their KRO undelegation
+     *         at the current block.
+     *
+     * @param validator The address of the validator.
+     * @param shares    The amount of shares to undelegate.
+     *
+     * @return The amount of assets that the Vault would exchange for the amount of shares provided.
+     */
+    function previewUndelegate(
+        address validator,
+        uint128 shares
+    ) public view virtual returns (uint128) {
+        return _convertToKroAssets(validator, shares);
+    }
+
+    /**
+     * @notice Allows an on-chain or off-chain user to simulate the effects of their KGH delegation
+     *         at the current block given current on-chain conditions.
+     *
+     * @param validator The address of the validator.
+     *
+     * @return The amount of shares that the Vault would exchange for the amount of assets provided.
+     */
+    function previewKghDelegate(address validator) public view virtual returns (uint128) {
+        return _convertToKghShares(validator, VKRO_PER_KGH);
+    }
+
+    /**
+     * @notice Allows an on-chain or off-chain user to simulate the effects of their KGH undelegation
+     *         at the current block, given current on-chain conditions.
+     *
+     * @param validator The address of the validator.
+     * @param tokenId   The tokenId of KGH to undelegate.
+     *
+     * @return The amount of assets that the Vault would exchange for the KGH of given tokenId.
+     */
+    function previewKghUndelegate(
+        address validator,
+        uint256 tokenId
+    ) public view virtual returns (uint128) {
+        return
+            _convertToKghAssets(
+                validator,
+                _vault[validator].kghDelegators[msg.sender].kghShares[tokenId]
+            );
+    }
+
+    /**
+     * @notice Returns the total amount of KRO assets held by the vault.
+     *
+     * @param validator Address of the validator.
+     *
+     * @return The total amount of KRO assets held by the vault.
+     */
+    function totalKroAssets(address validator) public view virtual returns (uint128) {
+        return _vault[validator].asset.totalKro;
+    }
+
+    /**
+     * @notice Returns the total amount of KGH assets held by the vault.
+     *
+     * @param validator Address of the validator.
+     *
+     * @return The total amount of KGH assets held by the vault.
+     */
+    function totalKghAssets(address validator) public view virtual returns (uint128) {
+        Vault storage vault = _vault[validator];
+        return vault.asset.totalKgh * VKRO_PER_KGH + vault.reward.boostedReward;
```

Comment:
Should `vault.asset.totalKroInKgh` be included in Kgh asset as well?

Reply:
KRO in KGH is handled using `kroShares`, so this function should exclude that value to calculate only boosted reward not base reward.
By the way, I think to change `total**` functions except for `totalKroAssets` to internal or private functions would be better. Because the return values of these functions include `VKRO_PER_KGH` or share offsets, so not appropriate for public purpose. I think `preview**` functions are enough for public use. Instead, how about remaining only `totalKroAssets` and adding `totalKghNum` function to return `_vault[validator].asset.totalKgh`?

Reply:
I agree to @seolaoh, that kind transition from `public` to `internal` functions will reduce the deployment cost.

Reply:
Thank you for your detailed explanation.
In other words, it is to distribute the boosted reward to all the KGH holders fairly? independent of the referral code? AFAIK, each KGH has different amount of KRO in KGH, but this is not considered in distribution of boosted reward. Do I understand, right?

BTW, I understood the point @seolaoh said and agreed with adding prefix of `_` for internal usage.

Reply:
Yes boosted reward is distributed uniformly to all KGH holders using `_convertToKghAssets`, and for KRO in KGH, the base reward will be distributed according to the number of KRO in KGH using `_convertToKroAssets`. So `totalKghAssets` should not include `totalKroInKgh`.

What I suggest is like below:
- convert `totalKghAssets`, `totalKroShares` and `totalKghShares` functions to internal or private functions
- add `totalKghNum` to return the total number of KGH delegated to the validator

Reply:
Good. I like it.